output.go

```diff
@@ -1,19 +1,80 @@
 package shaping
 
 import (
-	"golang.org/x/image/font/sfnt"
+	"github.com/benoitkugler/textlayout/harfbuzz"
+	"github.com/go-text/di"
 	"golang.org/x/image/math/fixed"
 )
 
-type Output interface {
-	// Advance returns the distance the Dot has advanced.
-	Advance() fixed.Int26_6
-	// Baseline returns the distance the baseline is from the top.
-	Baseline() fixed.Int26_6
-	// Bounds returns the smallest rectangle capable of containing the shaped text.
-	Bounds() fixed.Rectangle26_6
-	// Length returns the number of glyphs in the output.
-	Length() int
-	// Index returns the GlyphIndex to draw at the given index.
-	Index(int) sfnt.GlyphIndex
+type Glyph struct {
+	harfbuzz.GlyphInfo
+	harfbuzz.GlyphPosition
+}
+
+type Output struct {
+	// Advance is the distance the Dot has advanced.
+	Advance fixed.Int26_6
+	// Baseline is the distance the baseline is from the top.
+	Baseline fixed.Int26_6
+	// Bounds is the smallest rectangle capable of containing the shaped text.
+	Bounds fixed.Rectangle26_6
+	// Glyphs are the shaped output text.
+	Glyphs []Glyph
+}
+
+// Recalculate updates the Bounds, Advance, and Baseline fields in
+// the Output to match the current contents of the Glyphs field. This
+// operation requires querying extra information from the font as well
+// as knowing the direction of the text.
+func (o *Output) Recalculate(dir di.Direction, font *harfbuzz.Font) error {
```

Comment:
> for a simple LTR horizontal run, the advance, that is the Pos.XAdvance field, is always greater than the width (or horizontal extent), that is the GlyphExtents.Width field.

@benoitkugler Well, I think part of the issue is that I'm working with _complex_ scripts for my test cases. My work has all been using Arabic to exercise more of the shaper features.

What I've learned from Arabic is:

- Width is the visual width of the character. It has no fixed relationship with the XAdvance, as many Arabic characters have an XAdvance of 0 while still having a width.
- I also often have negative XBearing, because the glyph visually begins "before" the current position of the dot. Such glyphs often also have a right side bearing extending past the Xadvance.

I agree with both the diagram you linked and your math, but I encounter instances in which the advance does _not_ account for the bearings fairly often.

I guess that's why I'm confused. Since (in complex scripts) the character can overflow both sides of the `xadvance` pretty easily, how can pango use only the `xadvance` to wrap lines? It seems that like that would lead to jagged edges with strokes that hung past the edge of a bounding box.

It's also possible that there's some typographical convention that you don't need to align strokes that fall beyond Xadvance at the paragraph level. Maybe they're considered inconsequential, or even designed to hang over the edge? I don't really know.

I'm also unclear on whether the Glyph's XOffset should be considered as part of the width math or not. Do you happen to have any guidance there?

Reply:
I think the advance should still be enough. Here is why : quoting from Harfbuzz source, `Pos.XAdvance` is

>  How much the line advances **after** drawing this glyph when setting text in horizontal direction

and `Pos.XOffset` is  

> How much the glyph moves on the X-axis **before** drawing it, this should not affect how much the line advances.

So (ignoring Y coordinate for simplicity), the algorithm for drawing glyphs is the following (taken from pango) : 

1. Render the current glyph at (x + XOffset, y + YOffset), where (x, y) is the current point
2. Advance the current point to (x + XAdvance, y)
3. Render the next glyph...

Some glyphs may indeed have 0 as XAdvance, but I think that in this case, a previous glyph would already have incremented the advance, so that it would not be visually disturbing. 

Could you share an example of rune input (and font) you have tried ? I'm maybe over-simplifying here, I would like to check !

Reply:
![image](https://user-images.githubusercontent.com/3922125/140083985-80790a90-a384-49ff-ac90-3c4722c2c598.png)

This visualization is of a word (الأحلام) in Arabic using [this font](https://fonts.google.com/specimen/Scheherazade+New?subset=arabic#standard-styles). The red-ish line beneath each glyph is the Xadvance of that glyph. Each glyph stair-steps 4 pixels downwards so that it's easier to see the boundaries between them.


The green line represents the width of each character. It's starting X position is `dot+g.XBearing`, and its ending X position is `dot+g.XBearing+g.Width`

Here's a scaled-up version to make seeing the lines easier:
![arabic-lines-scaled](https://user-images.githubusercontent.com/3922125/140085015-5f8d5ab5-b773-4b73-b66c-a475818691a9.png)

I hadn't actually added this visualization until this conversation, so thanks for asking. It's helpful to see these numbers in action.

In general, it looks like you're right. The only time that the bearings do crazy things is when two glyphs are designed to overlap. That _shouldn't_ happen at a line break :crossed_fingers: , so we can probably safely ignore the extents when both computing the width of a line and when doing line breaking. That will also considerably simplify things.

Given all of this though, I find that I'm confused about exactly when the extents actually do matter. I don't need them when rendering the glyphs. Are they just points of reference, or what?

Reply:
> Given all of this though, I find that I'm confused about exactly when the extents actually do matter. I don't need them when rendering the glyphs. Are they just points of reference, or what?

Good question ! Looking at the Pango code, it seems it serves two purposes : 

- Horizontal extents is used instead of Harbuzz advance as a fallback (it seems deprecated though)
- Pango provides high-level functions to query the extents of a line, a run or even a cluster. These functions are **not** used to decide where to break lines, but they could probably be used by the user, for instance when selecting text. I guess that extent is visually better than a box built on relying only on glyph advances ?

To conclude on this point, do you think it would make sense to expose Recalculate as a separate method, but not to call it when shaping ? So that the extra computation may be avoided when, say, only line breaking is needed.

Reply:
Yeah, I think the sanest available choice is to surface a method that computes _just_ the advance width of an output. I find that I need to recompute it pretty often (after shaping) when doing line breaking, so making it a method on the output is still useful.

As for the extents, I agree that they aren't needed when line breaking, and I'm no longer convinced that the math I was doing with them is useful. However, I suspect they are needed when doing text selection. The tricky thing is, the only way to get accurate extent information is to re-configure the same font face at the same scale so that you can query for them in the same units as the rest of the dimensions.

Would it be reasonable to add the bearing/extent/dimension data to the `shaping.Glyph` type? Or do you think that it's too expensive to extract that data from the font to do it every time?

Reply:
Actually, we still need to query the extents in order to provide a tight bounding box on the glyphs. I think, for now, that we should include the extents of each glyph in the Glyph type in order to avoid higher-level code needing to re-query for that information.

Reply:
Alright !
I guess that it's OK for now not to worry too much on performances and handle it later if needed ?
On this point, I note that Pango caches the glyph extents. We can probably add a caching layer to the shaper if needed (something like a map[Face, size, Glyph]Extents)

Reply:
> Is this expected to be called by anyone other than the shaper? If not, exporting it pollutes the public API and it might be cleaner if the shaper calculates the values and set the fields so Output remains a 'dumb' struct

Did we get to the bottom of this original question? If it might be useful in the future can we leave it unexported for now?

Reply:
I think it may be useful to applications trying to get a tight bounding box on text after line breaking. Line breaking requires us to chop up the resulting output struct and recalculate parts of its dimensions, and I think providing helpers to do that is very reasonable. The RecalculateAdvance method is absolutely essential in the API for line breaking logic. We can leave the equivalent of this method unexported for now, but I expect that users will need it eventually.

Reply:
I'm not sure I completely follow - can you illustrate how it is used perhaps?
It seems very similar to updating the `Input` and calling `Shape` again, to understand the difference may help.

Reply:
You can break a line without reshaping if harfbuzz doesn't flag the boundary cluster as unsafe to break. This means that you can create a new line with something like:

```go
var out Output // assume we already have this from the shaper
newLine := out
newLine.Glyphs=newLine.Glyphs[start:end]
newLine.RecalculateAll()
```

Reply:
OK, that is kindof cool. But if we don't have to re-shape then what does recalculate mean?
It seems like there is power in here, but perhaps we can find a clearer way to express this in a future item of work (can it be unexported for now?)

Reply:
Harfbuzz only returns metrics for individual glyphs. The purpose of the `RecalculateAll()` method is to aggregate these metrics for the whole `Input` line (or run).
If we wan't to make this clearer, I think we could choose a "pure" approach : remove the line fields from Output (`Advance` and `GlyphBounds`), and make `RecalculateAll` return a struct containing them. 
(This will force the consumer to call it, but it will probably almost be the case anyway.)